### ISSUE TITLE
Found out the cause of possible reconnection infinite loops with proxies / bad connection + bug (?) with auto_reconnect argument [Explanation + proposed fixes]

### ISSUE THREAD

**Issue by vegeta1k95** [LonamiWebs/Telethon#1541]
I finally managed to get my hands on a SOCKS5 proxy that produce an infinite reconnection loop in the client and investigated the problem.

So, I used the following code setup for testing:

``` python
...
proxy = (...)
client = TelegramClient(...,
                                    timeout=2,
                                    proxy=proxy,
                                    request_retries=2,
                                    connection_retries=0,
                                    auto_reconnect=False)

await client.connect()
await client(functions.account.UpdateStatusRequest(offline=False))
await client.disconnect()
...
```

**The picture is as follows:** With the proxy, the code above stucks at the `await client.connect()` line and never reaches the next `UpdateStatus` request. With the time scripts start to consume more and more memory and log shown multiple warnings from asyncio telling about coroutines being cancelled or never resolved while pending.

**The cause: Long Story Short**  Either the proxy itself or Telegram servers **silently** (via proper peer shutdown) close the socket from their side just after the connection is established, leading to an attempt to auto reconnect by `recv_loop`
**(which actually shouldn't happen if either `auto_reconnect=False` or `connection_retries=0`)**, reconnection successes (!), but the server-side socket is again immediately silently closes, repeating the story and producting infinite loop of auto reconnection attempts.

**The cause: Explanation** 

1. The socket of the connection (in my case it was default `ConnectionTcpFull`) is **succesfully** connected  to the telegram server via the proxy, that is: the method `Connection._connect(...)` (in `connection.py`) succesfully creates socket, does the handshake and creates `SteamReader`, `StreamWriter` objects from `await asyncio.open_connection(sock=s)` line (!), so that `Connection._send_loop()` and  `Connection._recv_loop()` are started.

2. **For some strange reason**, the server-side socket (is it proxy or Telegram?) properly (silently) shutsdown the peer from it's side just immediately after we established our connection. The socket from our side is not properly shutdown yet and left in _send only_ mode (can only send bytes, but not receive). 

3. Because of the server-side peer shutdown the `recv_loop` gets 0 bytes packet immediately after the connection was established and raises exception in `FullPacketCodec.read_packet()` which is then caught in `_recv_loop()` of `MTProtoSender`. 

4. `MTProtoSender` then decides to do a `_start_reconnect()` attempt, which launches `_reconnect()`.

5. **In `MTProtoSender` in `_reconnect()`, in the lines**
``` python
    retries = self._retries if self._auto_reconnect else 0
    for attempt in retry_range(retries):
         ....
         await self._connect()
         ...
         if self._auto_reconnect_callback:
                asyncio.get_event_loop().create_task(self._auto_reconnect_callback())
```
**the condition on the `self._auto_reconnect` seems to be useless here (!!!), because the `retry_range()` function always yields `1` as** **the first value -> even if `_auto_reconnect=False` (hence `retries` is 0) the `retry_range` will** **spit out `1` and the `for` loop will** **not be skipped (!), leading to the attempt to connect again.**

6. Attempt to connect **succeds** again, but the server-side socket is again silently closed immediately after. `recv` and `send` loops are restarted, leading to the same repeating story from 1) and producting an infinite loop of reconnection attempts.

7. All this time, our main code is stuck on `await client.connect()` because we are stuck on the line:
``` python 
...
await self._sender.send(self._init_with(
            functions.help.GetConfigRequest()))
...
```
of the `TelegramBaseClient` in `connect()`.

**This is due to the fact, that the socket closure and the infinite reconnection** **loop prevent the future created by** **`MTProtoSender.send()` to being resolved neither to a** **sensible result (because `recv_loop` receives nothing) nor to an** **exception (because constant auto reconnection prevents** **futures being cancelled and exceptions are never set - e.g.** **`MTProtoSender._disconnect()` is never reached)**

8. Moreover, each such attempt to auto reconnect creates and places in queue a new request (with future) due to the `self._auto_reconnect_callback()` in the code line from item 5).  This is because each time auto reconnect happens the
coroutine `_handle_auto_reconnect()` from `UpdateMethods` puts a `GetUsersRequest` in the pending requests queue (`await self.get_me()`).
Thus, each iteration our pending states queue grows larger and larger, consuming more and more memory, but the states are never resolved to a response / exception.

9. This contniues until the proxy is down or some network abruption happens, thus might be for quite a long time.

That's it, this is what I have found.

**Suggestions:**

1. First of all, I think I would be better to fix the use of `auto_reconnect` parameter. I've looked up the older version of Telethon and there instead of 
``` python
retries = self._retries if self._auto_reconnect else 0
for attempt in retry_range(retires):
   ...
else:
  ...
```
was 
``` python
retries = self._retries if self._auto_reconnect else 0
for attempt in range(1, retries + 1):
   ...
else:
   ...
```
which produced more accurate behavior for `auto_reconnect=False` - for `retries=0` we skip the `for` loop straight into `else` condition which disconnects client completely. I think that what one would expect from `retries` parameter. (Why did we switched to a generator function in the first place?)

2. Maybe there should be a global limit to overall attempted reconnections, exceeding which, the client disconnects fully and/or raises and exception? 

3. To detect this silent closure of a connection from the server side, maybe we would want to ask a socket to `recv` 1 byte of data with timeout after the connection established, but before starting `send` and `recv` loop? 
E.g. in `connection.py` :
```python
...
await asyncio.wait_for(
                asyncio.get_event_loop().sock_connect(s, address),
                timeout=timeout
)

# Here we add some precautions to detect silent closure of the server-side socket.
# Most probably, before sending any connection initialization requests to Telegram, we suppose 
# to recieve nothing from the socket, but if receive empty bytes - the connection is already gone, 
# so we should check and do something about it before even staring reader/writer streams.
try:
    bytes_received = await asyncio.wait_for(
                    asyncio.get_event_loop().sock_recv(s, 1)¶,
                    timeout=timeout
    )
except asyncio.TimeoutError:
    # If timeout passes - everything is OK, we didn't received empty bytes
    pass
except Exception:
    # Some other network errors might have occured. Disconnect and do cleanup here.
    ...
else:   
   if not bytes_received:
       raise ConnectionError("Server closed the connection!")
...
```
If the socket receives empty packet (0 bytes) after the connection established, that means the proxy or Telegram immediately refused our connection and we might want to raise an exception.

4. Do we really need to automatically request `GetUsersRequest` after each reconnection to keep Telegram sending us updates?
Maybe we should let the users implement the logic of updates on the reconnection themselves?

5. Maybe we can try to wrap a call from 7) with `await asyncio.wait_for(...)` or similar, for some sensible timeout, and handle the exception?

There might be other possibly related issues, sticked to this one, which I overlooked.
Any suggestions and comments are welcome!

**Comment by Lonami** [comment]
Thanks for the investigation. There's a lot to unpack so I won't go in detail right now, but regarding your suggestions:

1. The helper method was added by https://github.com/LonamiWebs/Telethon/commit/c8f16a4e893e76d6fa4bf7acef4e1a3bce934caf to address inconsistencies in the code base.

2. Over long periods, this limit would probably always be reached. Also, the problem here is reconnection succeeds.

3. The server may not send anything at start, and if it did, that byte would be important to correctly read future data so it cannot simply be discarded.

4. The library should properly handle calls to `getDifference` as needed.

Is the proxy HTTP-only? It might close as soon as non-HTTP data is sent.

"Successful reconnect but closes short after" is definitely something that should count as "didn't reconnect fine", but it might be messy to add.

**Comment by Lonami** [comment]
I'm cleaning up proxy-related issues, so closing in favour of the new issue.
